service/history/workflow_task_handler_callbacks.go

```diff
@@ -606,6 +606,23 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 			return nil, err
 		}
 
+		if !ms.IsWorkflowExecutionRunning() {
+			// If workflow competed itself with one of the completion command, cancel all incomplete updates in the registry.
+			// Because all unprocessed updates were already rejected, incomplete updates in the registry are:
+			// - updates that were received while this WT was running,
+			// - updates that were accepted but not completed by this WT.
+			err = weContext.UpdateRegistry(ctx).CancelIncomplete(ctx, update.CancelReasonWorkflowCompleted, workflow.WithEffects(&effects, ms))
```

Comment:
Are updates rejected before or after we persist the mutable state change? If it's before, then the persistence could fail, and workflow remain open while updates are already rejected due to workflow completed.

Reply:
All operations on futures are buffered and applied only after MS is persisted. `effects` (last parameter) is responsible for it.

Reply:
ah, I see. I am still not familiar with the effect package enough.